### PR TITLE
[WIP] Truncated Normal Distribution

### PR DESCRIPTION
**Pull request by alicanb**
I patched up something consistent from old backups. No tests yet, will work on that shortly, but should give an idea about what to expect.

### CODE REVIEW

**Comment by alicanb** [comment]
cc @vincentqb @fritzo @neerajprad @vishwakftw @Enealor @cossio

**Comment by alicanb** [comment]
Still consistent with this notebook : https://gist.github.com/alicanb/c9e6567b7c512140ed43916b4dd30106

**Comment by cossio** [comment]
Thanks for this! It looks good.

On a tanential note, I was surprised to see that this is just Python code (I thought some C++ would be required for performance). Other distributions are implemented like this?

File: torch/distributions/truncated_normal.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    Available online: https://arxiv.org/abs/0907.4010
+
+    Args:
+        lower_bound (Tensor): lower bound for standard normal distribution. Best to keep it greater than -4.0 for
+        stable results
+        upper_bound (Tensor): upper bound for standard normal distribution. Best to keep it smaller than 4.0 for
+        stable results
+    """
+    x = torch.randn(sample_shape)
+    done = torch.zeros(sample_shape).byte()
```
Comment (vishwakftw) [inline torch/distributions/truncated_normal.py]: nit: `torch.zeros(sample_shape, dtype=torch.byte)`

File: torch/distributions/truncated_normal.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    while not done.all():
+        proposed_x = lower_bound + torch.rand(sample_shape) * (upper_bound - lower_bound)
+        if (upper_bound * lower_bound).lt(0.0):  # of opposite sign
+            log_prob_accept = -0.5 * proposed_x**2
+        elif upper_bound < 0.0:  # both negative
+            log_prob_accept = 0.5 * (upper_bound**2 - proposed_x**2)
+        else:  # both positive
+            assert(lower_bound.gt(0.0))
+            log_prob_accept = 0.5 * (lower_bound**2 - proposed_x**2)
+        prob_accept = torch.exp(log_prob_accept).clamp_(0.0, 1.0)
```
Comment (vishwakftw) [inline torch/distributions/truncated_normal.py]: Is the lower clamp required?

File: torch/distributions/truncated_normal.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        accept = torch.bernoulli(prob_accept).byte() & ~done
```
Comment (vishwakftw) [inline torch/distributions/truncated_normal.py]: `bool()`?

File: torch/distributions/truncated_normal.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import torch
+from torch.distributions.normal import Normal
+from torch.distributions.truncated_distribution import TruncatedDistribution
+import math
+
+
+def _standard_truncnorm_sample(lower_bound, upper_bound, sample_shape=torch.Size()):
```
Comment (fritzo) [inline torch/distributions/truncated_normal.py]: IIUC this is not reparameterized, and is only suitable as an implementation of `.sample()` (in particular it cannot differentiate wrt the bounds). I believe we can wrap this in a `torch.autograd.Function` and use the `erfinv` backward for gradients. Then it would also be suitable for `.rsample()`.

**Review by fritzo — CHANGES_REQUESTED** [review]

File: torch/distributions/truncated_normal.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    Args:
+        lower_bound (Tensor): lower bound for standard normal distribution. Best to keep it greater than -4.0 for
+        stable results
+        upper_bound (Tensor): upper bound for standard normal distribution. Best to keep it smaller than 4.0 for
+        stable results
+    """
+    x = torch.randn(sample_shape)
+    done = torch.zeros(sample_shape).byte()
+    while not done.all():
```
Comment (fritzo) [inline torch/distributions/truncated_normal.py]: @suo should we wrap this function in a `@torch.jit.script` so the `while` loop is correctly recorded even when `TruncatedNormal.sample()` appears inside a `torch.jit.trace`? I have started using this pattern in Pyro whenever possible.

File: torch/distributions/truncated_normal.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    Args:
+        loc (Tensor): mean of the distribution
+        scale (Tensor): standard deviation of the distribution
+        lower_bound (Tensor): lower bound for the distribution. Best to keep it greater than loc-4*scale for
+        stable results
+        upper_bound (Tensor): upper bound for standard normal distribution. Best to keep it smaller than loc+4*scale for
+        stable results
+    """
+
+    def __init__(self, loc, scale, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
```
Comment (fritzo) [inline torch/distributions/truncated_normal.py]: `*args, **kwargs` -> `validate_args=None`

File: torch/distributions/truncated_distribution.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to a base distribution.
+    """
+
+    has_rsample = True
+
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: `*args, **kwargs` -> `validate_args=None`

File: torch/distributions/truncated_distribution.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: I'm pretty sure we'll need to compute `batch_shape` and `event_shape` by broadcasting `base_distribution.batch_shape+base_distribution.event_shape` with `lower_bound` and `upper_bound` shapes.

File: torch/distributions/truncated_distribution.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        self.lower_bound, self.upper_bound = broadcast_all(lower_bound, upper_bound)
+        self.base_dist = base_distribution
+        cdf_low, cdf_high = self.base_dist.cdf(self.lower_bound), self.base_dist.cdf(self.upper_bound)
+        self.Z = (cdf_high - cdf_low)
+
+    @constraints.dependent_property
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: While this looks reasonable, I think after this PR we'll need logic in the transform registry to recognize half-lines as distinct from proper intervals, and use `ExpTransform` rather than `SigmoidTransform+Affine`. Something conceptually like
```py
MaskedTransform(lower_bound==float('inf'),
                MaskedTransform(upper_bound == float('inf'),
                                IdentityTransform(),
                                Afine(0,-1) + ExpTransform()),
                MaskedTransform(upper_bound == float('inf'),
                                ExpTransform(),
                                AffineTransform(...) + SigmoidTransform()))
```
cc @neerajprad

File: torch/distributions/truncated_normal.py (hunk 5 above)
Comment (alicanb) [reply torch/distributions/truncated_normal.py]: @fritzo is there an example i can take a look at?

Comment (fritzo) [reply torch/distributions/truncated_normal.py]: https://github.com/pyro-ppl/pyro/blob/82c7cfd9a93ef06663f303a85885a59912aa5f47/pyro/distributions/hmm.py#L15

**Review by V-Enzo — CHANGES_REQUESTED** [review]

**Comment by wallyxie** [comment]
@alicanb, thank you for tackling this. I've been getting up to date on the conversations surrounding the implementation of a `TruncatedNormal` distribution class in PyTorch after becoming involved in an applied inference project for which it would critical. I just wanted to follow up to see whether development on a CUDA-capable PyTorch truncated normal was still active or in hiatus? @toshas previously developed a PyTorch truncated normal module [here](https://github.com/toshas/torch_truncnorm/issues/2), though it has issues accepting `loc` and `scale` living on CUDA devices. Thanks again for all the time you've spent on this.

**Comment by wallyxie** [comment]
With the resolution of #https://github.com/toshas/torch_truncnorm/issues/2, it looks like there is now a fully functioning module for a `TruncatedNormal` distribution class. Thanks again, @toshas!

**Comment by fritzo** [comment]
bump

**Comment by majidaldo** [comment]
bump
